### PR TITLE
Add Custom Data to Web User Invite

### PR DESCRIPTION
**Pull request by Jtang-1**
## Product Description

Allows the user to define "user data" when creating a web user invite.

Screen Shot 2024-09-13 at 3 10 12 PM

## Technical Summary

[USH-4895](https://dimagi.atlassian.net/browse/USH-4895)
Editing custom user data uses the `CustomDataEditor` class which generates the profile and user data fields for `CustomDataForm`. The need here is very similar from the form UI side. However, it does not require populating existing user data. 

In order to reuse the `CustomDataForm`, I initialized an instance of `CustomDataEditor` which initializes the form. Then this instance of `CustomDataEditor` is injected into `AdminInvitesUserForm`. The fields and fieldsets created as part of `CustomDataForm ` is then pulled out to create the field and fieldsets for `AdminInvitesUserForm`. One issue faced is that `CustomDataForm` applies a form level defined prefix for each field but this prefix is applied only upon rendering the form. But since we are not using that form directly and are only extracting the fields, adding the prefix needs to be manually handled in `AdminInvitesUserForm`.

Additionally, when a user data field is defined and populated by the profile, the field is disabled. Because of this, posted data will not include data for this field. `CustomDataForm` already handles that and adds profile related fields to the post data. `AdminInvitesUserForm` takes advantage of that by accessing post data via `self.custom_data.form.data`.

## Feature Flag

New feature flag: `WEB_USER_INVITE_ADDITIONAL_FIELDS`

### Safety story

Locally tested and both backend and frontend changes are gated behind a FF which limits the blast radius.

### Automated test coverage

No automated test

### QA Plan

[QA-7109
](https://dimagi.atlassian.net/browse/QA-7109)

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

[USH-4895]: https://dimagi.atlassian.net/browse/USH-4895?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: corehq/toggles/__init__.py [hunk 1]
```diff
@@ -2528,6 +2528,13 @@ def _commtrackify(domain_name, toggle_is_enabled):
     namespaces=[NAMESPACE_DOMAIN]
 )
 
+WEB_USER_INVITE_ADDITIONAL_FIELDS = StaticToggle(
+    'web_user_invite_additional_fields',
+    'Enable additional fields in web user invite form for enhanced user details',
```
Comment (MartinRiese) [inline corehq/toggles/__init__.py:2533]: Should this have a "USH:" in front or is the automatically GAed?

File: corehq/apps/users/models.py [hunk 2]
```diff
@@ -552,9 +552,11 @@ def add_domain_membership(self, domain, timezone=None, **kwargs):
 
     def add_as_web_user(self, domain, role, primary_location_id=None,
                         assigned_location_ids=None, program_id=None, profile=None,
-                        tableau_role=None, tableau_group_ids=None):
+                        tableau_role=None, tableau_group_ids=None, custom_user_data=None):
         if assigned_location_ids is None:
             assigned_location_ids = []
+        if custom_user_data is None:
```
Comment (MartinRiese) [inline corehq/apps/users/models.py:558]: is this needed? It seems like the check on line 572 would take care of this and I don's see custom_user_data used anywhere else.

File: corehq/toggles/__init__.py (hunk 1 above)
Comment (Jtang-1) [reply corehq/toggles/__init__.py]: Yes it should [63273ea](https://github.com/dimagi/commcare-hq/pull/35124/commits/63273eab0379d42d39031a9a517efeb244ba4bf1)

File: corehq/apps/users/models.py (hunk 2 above)
Comment (Jtang-1) [reply corehq/apps/users/models.py]: you're right, that's not needed. [8405864](https://github.com/dimagi/commcare-hq/pull/35124/commits/840586437bdeea44cf652f2ef0cad18eeae92f0b)

File: corehq/apps/registration/forms.py [hunk 3]
```diff
@@ -490,23 +491,24 @@ class AdminInvitesUserForm(SelectUserLocationForm):
...
+            data = data.copy()
+            custom_data_post_dict = self.custom_data.form.data
+            data.update({k: v for k, v in custom_data_post_dict.items() if k not in data})
         self.request = kwargs.get('request')
         super(AdminInvitesUserForm, self).__init__(domain=domain, data=data, **kwargs)
         self.can_edit_tableau_config = can_edit_tableau_config
         domain_obj = Domain.get_by_name(domain)
         self.fields['role'].choices = [('', _("Select a role"))] + role_choices
         if domain_obj:
-            if domain_has_privilege(domain_obj.name, privileges.APP_USER_PROFILES):
```
Comment (minhaminha) [inline corehq/apps/registration/forms.py:500]: Might be missing something but why is this FF not being used anymore? Is it effectively putting all this behind the new flag?

File: corehq/apps/users/models.py [hunk 4]
```diff
@@ -564,9 +564,11 @@ def add_as_web_user(self, domain, role, primary_location_id=None,
             if primary_location_id:
                 self.set_location(domain, primary_location_id, commit=False)
             self.reset_locations(domain, assigned_location_ids, commit=False)
+        user_data = self.get_user_data(domain_obj.name)
         if domain_has_privilege(domain_obj.name, privileges.APP_USER_PROFILES) and profile:
-            user_data = self.get_user_data(domain_obj.name)
```
Comment (minhaminha) [inline corehq/apps/users/models.py:568]: also here

**Review by minhaminha — COMMENTED** [review]
May have to read through again but looks good so far.

File: corehq/apps/registration/forms.py (hunk 3 above)
Comment (Jtang-1) [reply corehq/apps/registration/forms.py:500]: It should still be behind the domain privilege check, but it's now done from the `CustomDataEditor` form [here](https://github.com/dimagi/commcare-hq/blob/0698b3dfcde9554bb4107545b0e9fe9d8de943e1/corehq/apps/custom_data_fields/edit_entity.py#L147) If that conditional is `false`, then the `profile` field [here](https://github.com/dimagi/commcare-hq/blob/0698b3dfcde9554bb4107545b0e9fe9d8de943e1/corehq/apps/custom_data_fields/edit_entity.py#L170) won't be added to the fields meaning that `self.custom_data.form.fields` will also not contain the profile field.

File: corehq/apps/users/models.py (hunk 4 above)
Comment (Jtang-1) [reply corehq/apps/users/models.py:568]: This does still require the domain privilege check for profiles and I believe it does do the check. Or maybe i'm misunderstanding your comment

**Comment by Jtang-1** [comment]
bump for review

**Review by MartinRiese — APPROVED** [review]

File: corehq/apps/registration/forms.py [hunk 5]
```diff
@@ -584,13 +584,12 @@ def __init__(self, data=None, excluded_emails=None, is_add_user=None,
             ),
         )
 
-    def clean_profile(self):
```
[the remark this answers is not in the document]
Comment (Jtang-1) [reply corehq/apps/registration/forms.py:582]: The `clean_<fieldname>()` method is a hook for performing custom validation or any cleaning that is specific to that particular attribute on a specific field in a Django form. Here, the field containing the profile will actually have the field name "<prefix>_profile" so `clean_profile` won't work. So I moved the cleaning to the general `clean()` and handle it via `_validate_profile()`

**Pull request merged**